File: db/fsharp/SalesManagement.Infrastructure/Migrations/Migration_20250106_015_CreateOrderTable.fs
namespace SalesManagement.Infrastructure.Migrations

open FluentMigrator

[<Migration(20250106015L)>]
type Migration_20250106_015_CreateOrderTable() =
    inherit Migration()

    override this.Up() =
        this.Create.Table("受注データ")
            .WithColumn("受注番号").AsString(10).NotNullable().PrimaryKey()
            .WithColumn("受注日").AsDateTime().NotNullable().WithDefault(SystemMethods.CurrentDateTime)
            .WithColumn("得意先コード").AsString(8).NotNullable()
            .WithColumn("得意先枝番").AsInt32().NotNullable().WithDefaultValue(0)
            .WithColumn("社員コード").AsString(10).NotNullable()
            .WithColumn("納期").AsDateTime().Nullable()
            .WithColumn("受注金額合計").AsInt32().NotNullable().WithDefaultValue(0)
            .WithColumn("消費税合計").AsInt32().NotNullable().WithDefaultValue(0)
            .WithColumn("伝票備考").AsString(1000).Nullable()
            .WithColumn("部門コード").AsString(10).NotNullable()
            .WithColumn("作成日時").AsDateTime().NotNullable().WithDefault(SystemMethods.CurrentDateTime)
            .WithColumn("作成者名").AsString(100).NotNullable()
            .WithColumn("更新日時").AsDateTime().NotNullable().WithDefault(SystemMethods.CurrentDateTime)
            .WithColumn("更新者名").AsString(100).NotNullable()
        |> ignore

        this.Create.ForeignKey("fk_order_customer")
            .FromTable("受注データ").ForeignColumns("得意先コード", "得意先枝番")
            .ToTable("顧客マスタ").PrimaryColumns("顧客コード", "顧客枝番")
        |> ignore

        this.Create.ForeignKey("fk_order_employee")
            .FromTable("受注データ").ForeignColumn("社員コード")
            .ToTable("社員マスタ").PrimaryColumn("社員コード")
        |> ignore

        // 部門コードへの外部キーは設定しない（部門マスタは複合主キー）

        this.Create.Index("idx_受注データ_得意先")
            .OnTable("受注データ")
            .OnColumn("得意先コード").Ascending()
            .OnColumn("得意先枝番").Ascending()
        |> ignore

        this.Create.Index("idx_受注データ_受注日")
            .OnTable("受注データ")
            .OnColumn("受注日").Descending()
        |> ignore

        this.Execute.Sql("COMMENT ON TABLE 受注データ IS '顧客からの受注情報を管理するヘッダ'")
        this.Execute.Sql("COMMENT ON COLUMN 受注データ.納期 IS '顧客が希望する納品日'")

    override this.Down() =
        this.Delete.Table("受注データ") |> ignore

File: db/csharp/SalesManagement.Infrastructure/Migrations/Migration_20250106_015_CreateOrderTable.cs
using FluentMigrator;

namespace SalesManagement.Infrastructure.Migrations
{
    [Migration(20250106015)]
    public class Migration20250106015CreateOrderTable : Migration
    {
        public override void Up()
        {
            Create.Table("受注データ")
                .WithColumn("受注番号").AsString(10).NotNullable().PrimaryKey()
                .WithColumn("受注日").AsDateTime().NotNullable().WithDefault(SystemMethods.CurrentDateTime)
                .WithColumn("部門コード").AsString(20).NotNullable()
                .WithColumn("開始日").AsDate().NotNullable()
                .WithColumn("顧客コード").AsString(8).NotNullable()
                .WithColumn("顧客枝番").AsInt32().Nullable()
                .WithColumn("社員コード").AsString(20).NotNullable()
                .WithColumn("希望納期").AsDateTime().Nullable()
                .WithColumn("客先注文番号").AsString(20).Nullable()
                .WithColumn("倉庫コード").AsString(3).NotNullable()
                .WithColumn("受注金額合計").AsInt32().NotNullable().WithDefaultValue(0)
                .WithColumn("消費税合計").AsInt32().NotNullable().WithDefaultValue(0)
                .WithColumn("備考").AsString(1000).Nullable()
                .WithColumn("作成日時").AsDateTime().NotNullable().WithDefault(SystemMethods.CurrentDateTime)
                .WithColumn("作成者名").AsString(100).Nullable()
                .WithColumn("更新日時").AsDateTime().NotNullable().WithDefault(SystemMethods.CurrentDateTime)
                .WithColumn("更新者名").AsString(100).Nullable();

            // 複合外部キーは Execute.Sql で実装
            Execute.Sql(@"
                ALTER TABLE 受注データ
                ADD CONSTRAINT fk_order_customer
                FOREIGN KEY (顧客コード, 顧客枝番)
                REFERENCES 顧客マスタ (顧客コード, 顧客枝番)
            ");

            Create.ForeignKey("fk_order_employee")
                .FromTable("受注データ").ForeignColumn("社員コード")
                .ToTable("社員マスタ").PrimaryColumn("社員コード");

            Execute.Sql(@"
                ALTER TABLE 受注データ
                ADD CONSTRAINT fk_order_department
                FOREIGN KEY (部門コード, 開始日)
                REFERENCES 部門マスタ (部門コード, 開始日)
            ");

            Create.Index("idx_受注データ_顧客")
                .OnTable("受注データ")
                .OnColumn("顧客コード").Ascending()
                .OnColumn("顧客枝番").Ascending();

            Create.Index("idx_受注データ_受注日")
                .OnTable("受注データ")
                .OnColumn("受注日");

            IfDatabase("Postgres").Execute.Sql("COMMENT ON TABLE 受注データ IS '顧客からの注文情報を管理するヘッダ'");
            IfDatabase("Postgres").Execute.Sql("COMMENT ON COLUMN 受注データ.受注番号 IS '受注を一意に識別する番号'");
            IfDatabase("Postgres").Execute.Sql("COMMENT ON COLUMN 受注データ.希望納期 IS '顧客が希望する納期'");
        }

        public override void Down()
        {
            Delete.Table("受注データ");
        }
    }
}
